packages/redeem_solver/src/service.ts

```diff
@@ -54,63 +54,107 @@ export class RedeemService {
 	}
 
 	private async processSolvedRedeem(req: RedeemRequestWithInputs) {
-		for (const input of req.inputs) {
-			if (input.sign_id) continue;
+		const client = this.getSuiClient(req.sui_network);
 
+		for (const input of req.inputs) {
 			try {
-				logger.info({
-					msg: "Requesting signature for input",
-					redeemId: req.redeem_id,
-					utxoId: input.utxo_id,
-					inputIdx: input.input_index,
-				});
-
-				const client = this.getSuiClient(req.sui_network);
-
-				const message = await client.getSigHash(
-					req.redeem_id,
-					input.input_index,
-					req.nbtc_pkg,
-					req.nbtc_contract,
-				);
-
-				const presignId = await client.createGlobalPresign();
-				const nbtcPublicSignature = await client.createUserSigMessage(
-					input.dwallet_id,
-					presignId,
-					message,
-				);
-
-				const signId = await client.requestInputSignature(
-					req.redeem_id,
-					input.input_index,
-					nbtcPublicSignature,
-					presignId,
-					req.nbtc_pkg,
-					req.nbtc_contract,
-				);
-
-				await this.storage.updateInputSignature(req.redeem_id, input.utxo_id, signId);
-				logger.info({
-					msg: "Requested signature",
-					redeemId: req.redeem_id,
-					utxoId: input.utxo_id,
-					signId: signId,
-				});
+				if (!input.sign_id) {
+					await this.ensureInputSigned(client, req, input);
+				} else if (input.sign_id && !input.verified) {
+					await this.ensureInputVerified(client, req, input);
```

Comment:
why do we handle 2 conditions in the same function?

Reply:
yes, but under the same function

Reply:
Handling Ika signatures should be done through events (listening on Ika events), rather than pooling.
Otherwise we risk paying for tx that fail because Ika signature is not ready.